### PR TITLE
Disable 3D widget by default. Enable 3D widget for Carving GUI when the data is 3D

### PR DESCRIPTION
**Pull request by wolny**
We need to pass the workflow class name to the VolumeEditor in order to be able to correctly initialise the 3D widget, i.e. we want the 3D widget to be visible only for workflows for certain workflows (currently only the carving workflow (?))

### CODE REVIEW

**Comment by wolny** [comment]
BTW, the tests are failing now cause they depend on the `VolumeEditor` changes I've made here: https://github.com/ilastik/volumina/pull/198

**Comment by akreshuk** [comment]
I would prefer to keep ilastik and volumina as separated as possible. Volumina shouldn't really know anything about ilastik workflows. I would add an argument like "add_3d_widget" to the volumeEditor that the workflow could pass when creating it, default to "off".

**Comment by wolny** [comment]
Thanks for the comment @akreshuk. Totally makes sense, as passing the workflow name to the `VolumeEditor` (Volumina) breaches the boundaries between the two project. I'll pass just the `add_3d_widget` flag to the `VolumeEditor` as you suggest, and do all the necessary computation of the flag in `ilastik`.

**Comment by wolny** [comment]
BTW I've made the necessary changes in `carving/preprocessingGui.py` and `carving/carvingGui.py` (those are the two entry points which initialise the `VolumeEditor`). 

I've added the `is_3d()` helper function to both `carving/opCarving` and `carving/opPreprocessing` for, since I didn't want to refactor them too much (Lazyflow's `Operator` is their common class their inherit from, but it doesn't seem like a good place to put the `is_3d()` function in there. We might as well extract `is_3d()` to some utility class/module in ilastik if you have any suggestions).

File: ilastik/applets/layerViewer/layerViewerGui.py [hunk 1]
```diff
@@ -76,12 +76,12 @@ class LayerViewerGui(with_metaclass(LayerViewerGuiMetaclass, QWidget)):
     """
     Implements an applet GUI whose central widget is a VolumeEditor
     and whose layer controls simply contains a layer list widget.
-    Intended to be used as a subclass for applet GUI objects.
+    Intended to be used as a superclass for applet GUI objects.
```
Comment (k-dominik) [inline ilastik/applets/layerViewer/layerViewerGui.py]: nice catch :)

File: ilastik/applets/labeling/labelingGui.py [hunk 2]
```diff
@@ -133,7 +133,8 @@ def __init__(self):
             # Slot that gives a list of label names
             self.labelNames = None # labelNames.value
 
-    def __init__(self, parentApplet, labelingSlots, topLevelOperatorView, drawerUiPath=None, rawInputSlot=None, crosshair=True):
+    def __init__(self, parentApplet, labelingSlots, topLevelOperatorView, drawerUiPath=None, rawInputSlot=None,
+                 crosshair=True, add_3d_widget=False):
```
Comment (k-dominik) [inline ilastik/applets/labeling/labelingGui.py]: with `add_3d_widget` I have the same suggestion as in volumina - as th ewidget is added anyhow and only visibility is altered. `3d_widget_visible` ?

File: ilastik/applets/layerViewer/layerViewerGui.py [hunk 3]
```diff
@@ -115,7 +115,8 @@ def stopAndCleanUp(self):
     ###########################################
     ###########################################
 
-    def __init__(self, parentApplet, topLevelOperatorView, additionalMonitoredSlots=[], centralWidgetOnly=False, crosshair=True):
+    def __init__(self, parentApplet, topLevelOperatorView, additionalMonitoredSlots=[], centralWidgetOnly=False,
+                 crosshair=True, add_3d_widget=False):
```
Comment (k-dominik) [inline ilastik/applets/layerViewer/layerViewerGui.py]: `3d_widget_visible` ?

File: ilastik/applets/layerViewer/layerViewerGui.py [hunk 4]
```diff
@@ -173,7 +174,8 @@ def __init__(self, parentApplet, topLevelOperatorView, additionalMonitoredSlots=
         self.saved_layer_visibilities = None
 
         self._initCentralUic()
-        self._initEditor(crosshair=crosshair)
+
+        self._initEditor(crosshair=crosshair, add_3d_widget=add_3d_widget)
```
Comment (k-dominik) [inline ilastik/applets/layerViewer/layerViewerGui.py]: `3d_widget_visible` ?

File: ilastik/applets/layerViewer/layerViewerGui.py [hunk 5]
```diff
@@ -615,11 +617,11 @@ def _initCentralUic(self):
         localDir = os.path.split(__file__)[0]
         uic.loadUi(localDir+"/centralWidget.ui", self)
 
-    def _initEditor(self, crosshair):
+    def _initEditor(self, crosshair, add_3d_widget):
```
Comment (k-dominik) [inline ilastik/applets/layerViewer/layerViewerGui.py]: `3d_widget_visible` ?

File: ilastik/applets/layerViewer/layerViewerGui.py [hunk 6]
```diff
@@ -615,11 +617,11 @@ def _initCentralUic(self):
...
         """
         Initialize the Volume Editor GUI.
         """
-        self.editor = VolumeEditor(self.layerstack, parent=self, crosshair=crosshair)
+        self.editor = VolumeEditor(self.layerstack, parent=self, crosshair=crosshair, add_3d_widget=add_3d_widget)
```
Comment (k-dominik) [inline ilastik/applets/layerViewer/layerViewerGui.py]: `3d_widget_visible` ?

File: ilastik/workflows/carving/carvingGui.py [hunk 7]
```diff
@@ -252,11 +252,17 @@ def makeColortable():
...
+        tagged_shape.update(self.topLevelOperatorView.InputData.meta.getTaggedShape())
+        is_3d = tagged_shape['x'] > 1 and tagged_shape['y'] > 1 and tagged_shape['z'] > 1
+        return is_3d
+
     def _after_init(self):
         super(CarvingGui, self)._after_init()
-        if self.render:self._toggleSegmentation3D()
-        
+        if self.render:
+            self._toggleSegmentation3D()
```
Comment (k-dominik) [inline ilastik/workflows/carving/carvingGui.py]: ! really nice one. Didn't understand the old line at first glance :D

File: ilastik/workflows/carving/opPreprocessing.py [hunk 8]
```diff
@@ -492,5 +492,3 @@ def reset(self):
         '''reset sigma and filter to values of last preprocess'''
         self.applet._gui.setSigma(self.initialSigma)
         self.applet._gui.setFilter(self.initialFilter)
-
```
Comment (k-dominik) [inline ilastik/workflows/carving/opPreprocessing.py]: Seems a bit unrelated :)

File: ilastik/workflows/carving/carvingGui.py (hunk 7 above)
Comment (wolny) [reply ilastik/workflows/carving/carvingGui.py]: 😆

File: ilastik/applets/labeling/labelingGui.py (hunk 2 above)
Comment (wolny) [reply ilastik/applets/labeling/labelingGui.py]: yeah, that name makes more sense. I'll change the name here and in Volumina, so it's consistent everywhere. Thanks!

Comment (wolny) [reply ilastik/applets/labeling/labelingGui.py]: changed the name to `is_3d_widget_visible`, cause var names cannot start with a number :) ✅

File: ilastik/applets/layerViewer/layerViewerGui.py (hunk 3 above)
Comment (wolny) [reply ilastik/applets/layerViewer/layerViewerGui.py]: ✅

File: ilastik/applets/layerViewer/layerViewerGui.py (hunk 4 above)
Comment (wolny) [reply ilastik/applets/layerViewer/layerViewerGui.py]: ✅

File: ilastik/applets/layerViewer/layerViewerGui.py (hunk 5 above)
Comment (wolny) [reply ilastik/applets/layerViewer/layerViewerGui.py]: ✅

File: ilastik/applets/layerViewer/layerViewerGui.py (hunk 6 above)
Comment (wolny) [reply ilastik/applets/layerViewer/layerViewerGui.py]: ✅

File: ilastik/workflows/carving/opPreprocessing.py (hunk 8 above)
Comment (wolny) [reply ilastik/workflows/carving/opPreprocessing.py]: ✅

**Comment by wolny** [comment]
Thanks for the review @k-dominik. I've changed the naming according to your suggestions, and removed the noisy import reformatting done by PyCharm.

**Comment by k-dominik** [comment]
Awesome, thank you for this very quick fix - already checked it out and it works :)
